nemo/collections/speechlm/models/base.py

```diff
@@ -0,0 +1,507 @@
+
+        self._num_test_dl = getattr(data_module, "_num_test_dl", None)
+        self._test_names = getattr(data_module, "_test_names", None)
+        if self._num_test_dl is None:
+            # special case for lhotse dataloader
+            self._num_test_dl = len(self._test_names)
+
+        if self._test_names is None:
+            if not self._num_test_dl:
+                raise ValueError(
+                    f"`_num_test_dl` not found/valid in data module: {getattr(data_module, '_num_test_dl', None)}"
+                )
+            self._test_names = [f'test_{idx}' for idx in range(self._num_test_dl)]
+            logging.info(f"`_test_names` not found in data module. Setting default names: {self._test_names}")
+        elif len(self._test_names) != self._num_test_dl:
+            raise ValueError(
+                f"Number of test names provided ({len(self._test_names)}) does not match number of test datasets ({self._num_test_dl})."
+            )
+
+    @property
+    def validation_step_outputs(self):
+        """
+        Cached outputs of validation_step. It can be a list of items (for single data loader) or a list of lists
+        (for multiple data loaders).
+
+        Returns:
+            List of outputs of validation_step.
+        """
+        if self._validation_step_outputs is not None:
+            return self._validation_step_outputs
+
+        # Initialize new output list
+        self._validation_step_outputs = []
+        # Check len(data_module._validation_ds) > 1 as sometimes single dataloader can be in a list: [<Dataloader obj>] when ds_item in
+        # config has 1 item passed in a list
+        if self._num_validation_dl > 1:
+            for _ in range(self._num_validation_dl):
+                self._validation_step_outputs.append([])
+
+        return self._validation_step_outputs
+
+    @validation_step_outputs.setter
+    def validation_step_outputs(self, value):
+        self._validation_step_outputs = value
+
+    @property
+    def test_step_outputs(self):
+        """
+        Cached outputs of test_step. It can be a list of items (for single data loader) or a list of lists (for multiple data loaders).
+
+        Returns:
+            List of outputs of test_step.
+        """
+        if self._test_step_outputs is not None:
+            return self._test_step_outputs
+
+        # Initialize new output list
+        self._test_step_outputs = []
+        # Check len(data_module._test_ds) > 1 as sometimes single dataloader can be in a list: [<Dataloader obj>] when ds_item in
+        # config has 1 item passed in a list
+        if self._num_test_dl > 1:
+            for _ in range(self._num_test_dl):
+                self._test_step_outputs.append([])
+
+        return self._test_step_outputs
+
+    @test_step_outputs.setter
+    def test_step_outputs(self, value):
+        self._test_step_outputs = value
+
+    def on_validation_epoch_end(self) -> Optional[Dict[str, Dict[str, torch.Tensor]]]:
```

Comment:
This.. can be refactored out of ModelPT as a mixin and shared between ModelPT and SpeechLanguageModel probably. But for now this is fine

Reply:
yeah I agree this can be a mixin, but leave it as it is for now to avoid more changes to ModelPT